File: connectors/connector-jira/src/main/java/com/taskadapter/connector/jira/JiraConnector.java
package com.taskadapter.connector.jira;

import com.atlassian.jira.rest.client.api.JiraRestClient;
import com.atlassian.jira.rest.client.api.domain.BasicComponent;
import com.atlassian.jira.rest.client.api.domain.Filter;
import com.atlassian.jira.rest.client.api.domain.IssueType;
import com.atlassian.jira.rest.client.api.domain.Version;
import com.taskadapter.connector.FieldRow;
import com.taskadapter.connector.NewConnector;
import com.taskadapter.connector.common.TaskSavingUtils;
import com.taskadapter.connector.definition.ProgressMonitor;
import com.taskadapter.connector.definition.SaveResult;
import com.taskadapter.connector.definition.TaskId;
import com.taskadapter.connector.definition.WebConnectorSetup;
import com.taskadapter.connector.definition.exceptions.BadConfigException;
import com.taskadapter.connector.definition.exceptions.ConnectorException;
import com.taskadapter.connector.definition.exceptions.ProjectNotSetException;
import com.taskadapter.core.PreviouslyCreatedTasksResolver;
import com.taskadapter.model.GTask;
import com.taskadapter.model.NamedKeyedObject;
import com.taskadapter.model.NamedKeyedObjectImpl;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

public class JiraConnector implements NewConnector {
    private static final Logger logger = LoggerFactory.getLogger(JiraConnector.class);

    /**
     * Keep it the same to enable backward compatibility with the existing
     * config files.
     */
    public static final String ID = "Atlassian JIRA";

    private final JiraConfig config;
    private final WebConnectorSetup setup;

    public JiraConnector(JiraConfig config, WebConnectorSetup setup) {
        this.config = config;
        this.setup = setup;
    }

    // XXX refactor this. we don't even need the IDs!

    public List<NamedKeyedObject> getFilters() throws ConnectorException {
        return withJiraRestClient(client -> {
            // TODO need all filters, not just favorites - but JIRA REST API does not support. (Dec 6 2015)
            var filtersPromise = client.getSearchClient().getFavouriteFilters();
            var filters = filtersPromise.claim();
            var list = new ArrayList<NamedKeyedObject>();
            for (Filter filter : filters) {
                list.add(new NamedKeyedObjectImpl(filter.getId() + "", filter.getName()));
            }
            return list;
        });
    }

    List<NamedKeyedObject> getComponents() throws ConnectorException {
        return withJiraRestClient(client -> {
            var projectKey = config.getProjectKey();
            if (projectKey == null) {
                throw new ProjectNotSetException();
            }
            var projectPromise = client.getProjectClient().getProject(projectKey);
            var project = projectPromise.claim();
            var components = project.getComponents();
            var list = new ArrayList<NamedKeyedObject>();
            for (BasicComponent c : components) {
                list.add(new NamedKeyedObjectImpl(String.valueOf(c.getId()), c.getName()));
            }
            return list;
        });
    }

    List<NamedKeyedObject> getVersions() throws ConnectorException {
        return withJiraRestClient(client -> {
            var projectKey = config.getProjectKey();
            if (projectKey == null) {
                throw new ProjectNotSetException();
            }
            var projectPromise = client.getProjectClient().getProject(projectKey);
            var project = projectPromise.claim();
            var versions = project.getVersions();
            var list = new ArrayList<NamedKeyedObject>();
            for (Version version : versions) {
                list.add(new NamedKeyedObjectImpl(String.valueOf(version.getId()), version.getName()));
            }
            return list;
        });
    }

    List<NamedKeyedObject> getAllIssueTypes() throws ConnectorException {
        return IssueTypesLoader.getIssueTypes(setup, new AllIssueTypesFilter());
    }

    List<? extends NamedKeyedObject> getIssueTypesForSubtasks() throws ConnectorException {
        return IssueTypesLoader.getIssueTypes(setup, new SubtaskTypesFilter());
    }

    public GTask loadTaskByKey(TaskId key, Iterable<FieldRow<?>> rows) throws ConnectorException {
        return withJiraRestClient(client -> {
            var loader = new JiraTaskLoader(client, config.getPriorities());
            return loader.loadTask(key.getKey());
        });
    }

    @Override
    public List<GTask> loadData() throws ConnectorException {
        return withJiraRestClient(client -> {
                    var loader = new JiraTaskLoader(client, config.getPriorities());
                    return loader.loadTasks(config);
                }
        );
    }

    @Override
    public SaveResult saveData(PreviouslyCreatedTasksResolver previouslyCreatedTasks, List<GTask> tasks,
                               ProgressMonitor monitor,
                               Iterable<FieldRow<?>> rows) {
        try {
            return withJiraRestClient(client -> {
                        var issueTypeList = loadIssueTypes(client);
                        var projectPromise = client.getProjectClient().getProject(config.getProjectKey());
                        var project = projectPromise.claim();
                        var versions = project.getVersions();
                        var components = project.getComponents();
                        // Need to load Jira server priorities because what we store in the config files is a
                        // priority name (string), while Jira returns the number value of the issue priority
                        var prioritiesPromise = client.getMetadataClient().getPriorities();
                        var priorities = prioritiesPromise.claim();

                        var resolver = JiraClientHelper.loadCustomFields(client);
                        var converter = new GTaskToJira(config, resolver, versions, components, priorities);
                        var saver = new JiraTaskSaver(client, issueTypeList, config.getDefaultTaskType(),
                                config.getDefaultIssueTypeForSubtasks());
                        var rb = TaskSavingUtils.saveTasks(previouslyCreatedTasks, tasks, converter, saver, monitor, rows,
                                setup.getHost());
                        TaskSavingUtils.saveRemappedRelations(config, tasks, saver, rb);
                        return rb.getResult();
                    }
            );
        } catch (ConnectorException e) {
            return new SaveResult("", 0, 0, List.of(), List.of(e), List.of());
        }
    }

    private Iterable<IssueType> loadIssueTypes(JiraRestClient jiraRestClient) throws BadConfigException {
        var issueTypeListPromise = jiraRestClient.getMetadataClient().getIssueTypes();
        var issueTypeList = issueTypeListPromise.claim();
        //check if default issue type exists in Jira
        for (IssueType anIssueTypeList : issueTypeList) {
            if (anIssueTypeList.getName().equals(config.getDefaultTaskType())) {
                return issueTypeList;
            }
        }
        throw new BadConfigException("Default issue type " + config.getDefaultTaskType() + " does not exist in JIRA");
    }

    private <T> T withJiraRestClient(JiraRestClientAction<T> f) throws ConnectorException {
        try (JiraRestClient client = JiraConnectionFactory.createClient(setup)) {
            return f.apply(client);
        } catch (Exception e) {
            throw JiraUtils.convertException(e);
        }
    }

    @FunctionalInterface
    interface JiraRestClientAction<T> {
        T apply(JiraRestClient client) throws IOException, ConnectorException;
    }
}

File: connectors/connector-jira/src/main/scala/com/taskadapter/connector/jira/JiraConnector.scala
package com.taskadapter.connector.jira

import java.io.IOException
import java.util

import com.atlassian.jira.rest.client.api.JiraRestClient
import com.atlassian.jira.rest.client.api.domain.IssueType
import com.taskadapter.connector.common.TaskSavingUtils
import com.taskadapter.connector.definition.exceptions.{BadConfigException, ConnectorException, ProjectNotSetException}
import com.taskadapter.connector.definition.{ProgressMonitor, SaveResult, TaskId, WebServerInfo}
import com.taskadapter.connector.{FieldRow, NewConnector}
import com.taskadapter.core.PreviouslyCreatedTasksResolver
import com.taskadapter.model.{GTask, NamedKeyedObject, NamedKeyedObjectImpl}
import org.slf4j.LoggerFactory


object JiraConnector {
  private val logger = LoggerFactory.getLogger(classOf[JiraConnector])
  /**
    * Keep it the same to enable backward compatibility with the existing
    * config files.
    */
  val ID = "Atlassian Jira"

  @FunctionalInterface private[jira] trait JiraRestClientAction[T] {
    @throws[IOException]
    @throws[ConnectorException]
    def apply(client: JiraRestClient): T
  }

}

class JiraConnector(val config: JiraConfig, var webServerInfo: WebServerInfo) extends NewConnector {
  // XXX refactor this. we don't even need the IDs!
  @throws[ConnectorException]
  def getFilters: util.List[NamedKeyedObject] = withJiraRestClient((client: JiraRestClient) => {
    def foo(client: JiraRestClient): util.List[NamedKeyedObject] = {
      // TODO need all filters, not just favorites - but JIRA REST API does not support. (Dec 6 2015)
      val filtersPromise = client.getSearchClient.getFavouriteFilters
      val filters = filtersPromise.claim
      val list = new util.ArrayList[NamedKeyedObject]
      import scala.collection.JavaConversions._
      for (filter <- filters) {
        list.add(new NamedKeyedObjectImpl(filter.getId + "", filter.getName))
      }
      list
    }

    foo(client)
  })

  @throws[ConnectorException]
  def getComponents: util.List[NamedKeyedObject] = withJiraRestClient((client: JiraRestClient) => {
    def foo(client: JiraRestClient) = {
      val projectKey = config.getProjectKey
      if (projectKey == null) throw new ProjectNotSetException
      val projectPromise = client.getProjectClient.getProject(projectKey)
      val project = projectPromise.claim
      val components = project.getComponents
      val list = new util.ArrayList[NamedKeyedObject]
      import scala.collection.JavaConversions._
      for (c <- components) {
        list.add(new NamedKeyedObjectImpl(String.valueOf(c.getId), c.getName))
      }
      list
    }

    foo(client)
  })

  @throws[ConnectorException]
  def getVersions: util.List[NamedKeyedObject] = withJiraRestClient((client: JiraRestClient) => {
    def foo(client: JiraRestClient) = {
      val projectKey = config.getProjectKey
      if (projectKey == null) throw new ProjectNotSetException
      val projectPromise = client.getProjectClient.getProject(projectKey)
      val project = projectPromise.claim
      val versions = project.getVersions
      val list = new util.ArrayList[NamedKeyedObject]
      import scala.collection.JavaConversions._
      for (version <- versions) {
        list.add(new NamedKeyedObjectImpl(String.valueOf(version.getId), version.getName))
      }
      list
    }

    foo(client)
  })

  @throws[ConnectorException]
  def getAllIssueTypes: util.List[NamedKeyedObject] = IssueTypesLoader.getIssueTypes(webServerInfo, new AllIssueTypesFilter)

  @throws[ConnectorException]
  def getIssueTypesForSubtasks: util.List[_ <: NamedKeyedObject] = IssueTypesLoader.getIssueTypes(webServerInfo, new SubtaskTypesFilter)

  override def loadTaskByKey(key: TaskId, rows: java.lang.Iterable[FieldRow]): GTask = withJiraRestClient((client: JiraRestClient) => {
    def foo(client: JiraRestClient) = {
      val loader = new JiraTaskLoader(client, config.getPriorities)
      loader.loadTask(key.key)
    }

    foo(client)
  })

  override def loadData: util.List[GTask] = withJiraRestClient((client: JiraRestClient) => {
    def foo(client: JiraRestClient) = {
      val loader = new JiraTaskLoader(client, config.getPriorities)
      loader.loadTasks(config)
    }

    foo(client)
  })

  override def saveData(previouslyCreatedTasks: PreviouslyCreatedTasksResolver, tasks: util.List[GTask], monitor: ProgressMonitor,
               rows: java.lang.Iterable[FieldRow]): SaveResult =
    withJiraRestClient((client: JiraRestClient) => {
    def foo(client: JiraRestClient) = {
      val issueTypeList = loadIssueTypes(client)
      val projectPromise = client.getProjectClient.getProject(config.getProjectKey)
      val project = projectPromise.claim
      val versions = project.getVersions
      val components = project.getComponents
      /* Need to load Jira server priorities because what we store in the config files is a
                   * priority name (string), while Jira returns the number value of the issue priority */ val prioritiesPromise = client.getMetadataClient.getPriorities
      val priorities = prioritiesPromise.claim
      val converter = new GTaskToJira(config, issueTypeList, versions, components, priorities)
      val saver = new JiraTaskSaver(client)
      val rb = TaskSavingUtils.saveTasks(previouslyCreatedTasks, tasks, converter, saver, monitor, rows)
      TaskSavingUtils.saveRemappedRelations(config, tasks, saver, rb)
      rb.getResult
    }

    foo(client)
  })

  @throws[BadConfigException]
  private def loadIssueTypes(jiraRestClient: JiraRestClient) : java.lang.Iterable[IssueType] = {
    val issueTypeListPromise = jiraRestClient.getMetadataClient.getIssueTypes
    val issueTypeList = issueTypeListPromise.claim
    //check if default issue type exists in Jira
    import scala.collection.JavaConversions._
    for (anIssueTypeList <- issueTypeList) {
      if (anIssueTypeList.getName == config.getDefaultTaskType) return issueTypeList
    }
    throw new BadConfigException("Default issue type " + config.getDefaultTaskType + " does not exist in JIRA")
  }

  private def withJiraRestClient[T](f: JiraConnector.JiraRestClientAction[T]) = try {
    val client = JiraConnectionFactory.createClient(webServerInfo)
    try
      f.apply(client)
    catch {
      case e: Exception =>
        //            throw JiraUtils.convertException(e);
        throw new RuntimeException(e)
    } finally if (client != null) client.close()
  }
}